esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -5885,4 +5885,182 @@ int CWsDfuEx::GetIndexData(IEspContext &context, bool bSchemaOnly, const char* i
+        if (nodeGroup)
+            partsOnCluster.setCanReplicate(nodeGroup->queryCanReplicate());
+
+        const char *defaultDir = fileDesc->queryDefaultDir();
+        if (!isEmptyString(defaultDir))
+        {
+            DFD_OS os = SepCharBaseOs(getPathSepChar(defaultDir));
+            StringBuffer baseDir, repDir;
+            clusterInfo->getBaseDir(baseDir, os);
+            clusterInfo->getReplicateDir(repDir, os);
+            partsOnCluster.setBaseDir(baseDir.str());
+            partsOnCluster.setReplicateDir(repDir.str());
+        }
+    }
+}
+
+bool CWsDfuEx::onDFUCreateAndPublish(IEspContext &context, IEspDFUCreateAndPublishRequest &req, IEspDFUCreateAndPublishResponse &resp)
+{
+    try
+    {
+        if (!context.validateFeatureAccess(FEATURE_URL, SecAccess_Write, false))
+            throw MakeStringException(ECLWATCH_DFU_ACCESS_DENIED, "Failed to CreateAndPublish. Permission denied.");
+
+        const char *logicalName = req.getLogicalName();
+        const char *groupName = req.getGroupName();
+        const char *jobID = req.getJobID();
+        const char *recordDefinition = req.getECLRecordDefinition();
+        if (isEmptyString(logicalName))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT,"No LogicalName defined.");
+        if (isEmptyString(jobID))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT,"No JobID defined.");
+        if (isEmptyString(recordDefinition))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT,"No ECLRecordDefinition defined.");
+        if (isEmptyString(groupName))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT,"No GroupName defined.");
+
+        StringBuffer userID;
+        context.getUserID(userID);
+
+        Owned<IUserDescriptor> userDesc;
+        if(!userID.isEmpty())
+        {
+            userDesc.setown(createUserDescriptor());
+            userDesc->set(userID.str(), context.queryPassword(), context.querySignature());
+        }
+
+        { //Throw an exception if the file exists inside the group.
+            Owned<IDistributedFile> df = queryDistributedFileDirectory().lookup(logicalName, userDesc); // lock super-owners
+            if(df)
+            {
+                unsigned numClusters = df->numClusters();
+                for (unsigned i = 0; i < numClusters; i++)
+                {
+                    StringBuffer gName;
+                    df->getClusterGroupName(i, gName);
+                    if (strieq(groupName, gName.str()))
+                        throw MakeStringException(ECLWATCH_INVALID_INPUT, "Logical file %s exsits in %s.", logicalName, groupName);
+                }
+            }
+        }
+
+        createAndPublishLogicalFile(context, logicalName, groupName, jobID, userID.str(), recordDefinition, userDesc);
+
+        IEspDFUFilePartsOnCluster &partsOnCluster = resp.updateParts();
+        getFilePartsOnCluster(context, logicalName, groupName, userDesc, partsOnCluster);
```

Comment:
as mentioned elsewhere, could this new service not be combined with DFUInfo ?
Isn't essentially what is happening here is a DFUInfo reply having 1st created a new file?